Wikibase/Sniffs/Commenting/ClassLevelDocumentationSniff.php

```diff
@@ -60,12 +62,31 @@ public function process( File $phpcsFile, $stackPtr ) {
 			return;
 		}
 
+		if ( $this->license ) {
+			$docClose = $phpcsFile->findPrevious( T_DOC_COMMENT_CLOSE_TAG, $stackPtr );
+			$docStart = $phpcsFile->findPrevious( T_DOC_COMMENT_OPEN_TAG, $docClose );
+			$docBlock = $phpcsFile->getTokensAsString( $docStart, $docClose );
+
+			if ( strpos( $docBlock, "@license {$this->license}" ) === false ) {
```

Comment:
Can we please, please create a new sniff for this check?
* It is not closely related to what the "class level documentation" sniff is currently meant to do: just check if there *is* documentation, but don't care what's inside. Let's please stick to this clean paradigm.
* I would like to use (and possibly disable) these sniffs separately.
* The current code is already somewhat complex. Let's please try to keep the complexity (per sniff) low.
* The new functionality does not really depend on any other code in this sniff.
* The code can become surprisingly trivial then: Register the new sniff for `T_DOC_COMMENT_TAG` and bail out on any tag that's not `@license`. Find the next `T_DOC_COMMENT_STRING` after `@license`. Check if this string is equal to the one you expect.
* This also avoids the (possibly) expensive getTokensAsStrings call.

Reply:
re having a separate sniff. I'd be in favour of this as well. Preferably I'd go with a sniff checking (and adding if missing) the license tag, and another one that checks that test classes have covers tag. The existing check for any class-level comment is virtually useful in my eyes, really. What @bekh6ex does here, I actually got as the first step towards replacing ClassLevelDocumentation sniff with the one for class-level license tags. So that would be a fine first step for me.

> The code can become surprisingly trivial then: Register the new sniff for T_DOC_COMMENT_TAG and bail out on any tag that's not @license. Find the next T_DOC_COMMENT_STRING after @license. Check if this string is equal to the one you expect.

This is not really that simple I think:
1. What about the case when there is no class-level comment at all? I'd expect this sniff to find and fix this as well. One could argue that the other sniff could take care of checking that class-level comments exist, but I think sniff should rather independent.
2. I assume (but I might be wrong) the intention would be to check class-level @license definitions only, i.e. not check/add license tags on any possible level. So the suggested idea would do some redundant check, or would require a logic check is the comment coming right before the class/interface/trait.

So the sniff registered for class etc definintion, and check the preceding comment would be a way to go IMO.

Not aware that @bekh6ex went ahead, I've been tinkering something along the lines of this idea. Not finished, less compact than the one here, no autofixing yet etc but you could have a look instead of reading all the words at https://github.com/wmde/WikibaseCodeSniffer/commit/0f821e6ecbf10d9aeda8c8b8321c1cf1f40073df. That other version is definitely not better, but just mentioning it here, so I could remove the code from my computer :)